### PR TITLE
Tweak `Symbol` and `InternedString`

### PR DESCRIPTION
**Pull request by nnethercote**
Some minor improvements to speed and code cleanliness.

r? @Zoxc

### CODE REVIEW

**Comment by Zoxc** [comment]
@rust-timer build 9a6dab6df27e156738a99f952fa5d5b40133be7d

**Comment by nnethercote** [comment]
I got miniscule improvements locally, but I think it's worthwhile from the clean-up angle.

File: src/libsyntax_pos/symbol.rs [hunk 1]
```diff
@@ -510,10 +515,19 @@ impl Interner {
         symbol.0.as_usize() >= self.strings.len()
     }
 
-    pub fn get(&self, symbol: Symbol) -> &str {
+    /// Get the chars of a normal (non-gensym) symbol.
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: The distinction between gensymed and non-gensymed should never leak into a public interface.
For the `Symbol` users gensyms should look just like usual names (that happen to have some unique hashes appended).
`is_gensymed` is unnecessary after https://github.com/rust-lang/rust/pull/51492 and every its use is a mistake.

Comment (petrochenkov) [reply src/libsyntax_pos/symbol.rs]: Gensyms are not a different "kind" of symbols.

File: src/libsyntax_pos/symbol.rs [hunk 2]
```diff
@@ -611,11 +625,17 @@ fn with_interner<T, F: FnOnce(&mut Interner) -> T>(f: F) -> T {
     GLOBALS.with(|globals| f(&mut *globals.symbol_interner.lock()))
 }
 
-/// Represents a string stored in the interner. Because the interner outlives any thread
-/// which uses this type, we can safely treat `string` which points to interner data,
-/// as an immortal string, as long as this type never crosses between threads.
-// FIXME: ensure that the interner outlives any thread which uses `LocalInternedString`,
-// by creating a new thread right after constructing the interner.
+/// An alternative to `Symbol` and `LocalInternedString`, useful when the chars
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]:
```suggestion
/// An alternative to `Symbol` and `InternedString`, useful when the chars
```

File: src/libsyntax_pos/symbol.rs [hunk 3]
```diff
@@ -611,11 +625,17 @@ fn with_interner<T, F: FnOnce(&mut Interner) -> T>(f: F) -> T {
...
+/// within the symbol need to be accessed. It is best used for temporary
+/// values.
+///
+/// Because the interner outlives any thread which uses this type, we can
+/// safely treat `string` which points to interner data, as an immortal string,
+/// as long as this type never crosses between threads.
+//
+// FIXME: ensure that the interner outlives any thread which uses
+// `LocalInternedString`, by creating a new thread right after constructing the
+// interner.
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: @Zoxc , I don't think this FIXME is necessary?

File: src/libsyntax_pos/symbol.rs [hunk 4]
```diff
@@ -708,7 +728,12 @@ impl Encodable for LocalInternedString {
     }
 }
 
-/// Represents a string stored in the string interner.
+/// A thin wrapper around `Symbol`. It has two main differences to `Symbol`.
+/// - Its implementations of `Hash`, `PartialOrd` and `Ord` work with the
+///   string chars rather than the symbol integer. This is useful when
+///   hash stability is required across compile sessions, or a guaranteed sort
+///   ordering is required.
+/// - It is guaranteed to not be a gensym symbol.
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: This guarantee doesn't make much sense, `InternedString` represents *string contents*, internal `Symbol` is an implementation detail.
`gensym_symbol.as_interned_str()` just loses its gensymness in the process of conversion because it's not a part of string contents (which causes issues, https://github.com/rust-lang/rust/issues/49300).

File: src/libsyntax_pos/symbol.rs [hunk 5]
```diff
@@ -717,14 +742,23 @@ pub struct InternedString {
...
-            interner.get(self.symbol) as *const str
+            interner.symbol_str(self.symbol) as *const str
         });
         // This is safe because the interner keeps string alive until it is dropped.
         // We can access it because we know the interner is still alive since we use a
         // scoped thread local to access it, and it was alive at the beginning of this scope
         unsafe { f(&*str) }
     }
 
+    pub fn with2<F: FnOnce(&str, &str) -> R, R>(self, other: &InternedString, f: F) -> R {
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]:
```suggestion
    fn with2<F: FnOnce(&str, &str) -> R, R>(self, other: &InternedString, f: F) -> R {
```

**Comment by petrochenkov** [comment]
Could you revert the `fn get` -> `fn symbol_str`/`fn symbol_or_gensym_str` change?
It's ideologically wrong and it's not faster since `symbol_str` is still doing checked indexing.

File: src/libsyntax_pos/symbol.rs (hunk 1 above)
Comment (nnethercote) [reply src/libsyntax_pos/symbol.rs]: Ok, I will fix.

File: src/libsyntax_pos/symbol.rs (hunk 4 above)
Comment (nnethercote) [reply src/libsyntax_pos/symbol.rs]: Ok, I will fix.

**Comment by nnethercote** [comment]
@petrochenkov: New code is up.

File: src/libsyntax_pos/symbol.rs (hunk 3 above)
Comment (Zoxc) [reply src/libsyntax_pos/symbol.rs]: Why not?

File: src/libsyntax_pos/symbol.rs [hunk 6]
```diff
@@ -367,10 +380,6 @@ impl Symbol {
         with_interner(|interner| interner.intern(string))
     }
 
-    pub fn interned(self) -> Self {
```
Comment (Zoxc) [inline src/libsyntax_pos/symbol.rs]: Why is this removed?

File: src/libsyntax_pos/symbol.rs (hunk 3 above)
Comment (petrochenkov) [reply src/libsyntax_pos/symbol.rs]: I assumed this already true due to how `GLOBALS` and threads are created.

File: src/libsyntax_pos/symbol.rs (hunk 6 above)
Comment (petrochenkov) [reply src/libsyntax_pos/symbol.rs]: No longer used.

**Comment by petrochenkov** [comment]
r=me is @Zoxc doesn't have more concerns

r? @Zoxc

File: src/libsyntax_pos/symbol.rs (hunk 3 above)
Comment (Zoxc) [reply src/libsyntax_pos/symbol.rs]: I think it happens do be done that way now, but it's not enforced or documented as required by `symbol.rs`.
